### PR TITLE
FXIOS-844 ⁃ Widget telemetry

### PR DESCRIPTION
**Pull request by sblatz**
*Thank you for submitting a pull request, your contributions are greatly appreciated!*

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Please make sure you've run the test scheme and all the tests pass. If your changes affect existing tests please make sure to update the tests as well.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/FXIOS-844)

### CODE REVIEW

**Comment by sblatz** [comment]
@liuche for data review.

# Request for data collection review form 
1. What questions will you answer with this data?
* Which widgets do users use most frequently?

2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:
* Provide information essential for advancing a business objective such as supporting OKRs.
* Help us prioritize feature work / bug fixes on widgets

3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
* No other way to collect this data.

4. Can current instrumentation answer these questions?
* No, widgets were just added and we have no existing telemetry for them.

5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki.

Measurement Description | Data Collection Category | Tracking Bug # |
| --- | ----- | ----- |
When the quick sites "search" widget is opened | 2 | https://github.com/mozilla-mobile/firefox-ios/issues/7097
When the quick sites "open link" widget is opened | 2 | https://github.com/mozilla-mobile/firefox-ios/issues/7097
When the quick sites "close private tabs" widget is opened | 2 | https://github.com/mozilla-mobile/firefox-ios/issues/7097
When the quick sites "search privately" widget is opened | 2 | https://github.com/mozilla-mobile/firefox-ios/issues/7097
When the quick sites "open tabs" widget is opened | 2 | https://github.com/mozilla-mobile/firefox-ios/issues/7097
When the quick sites "top sites" widget is opened | 2 | https://github.com/mozilla-mobile/firefox-ios/issues/7097

6. How long will this data be collected? Choose one of the following:
* I want this data to be collected for 6 months initially (potentially renewable).

7. What populations will you measure?
All countries, locales, and channels that have data collection on.

8. If this data collection is default on, what is the opt-out mechanism for users?
Toggle switch for 'Send Usage Data' in the app settings.

9. Please provide a general description of how you will analyze this data.
By monitoring the count of how often users are performing these actions.

10. Where do you intend to share the results of your analysis?
With the mobile team.

11. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:
No.

**Review by liuche — CHANGES_REQUESTED** [review]
I don't see any metrics.yaml changes, so does this change not use Glean for telemetry?

One thing that I'll need before doing data review is public documentation for this (such as what metrics.yaml does - maybe iOS has a separate non-glean telemetry doc that you could add to, if this isn't going to use Glean?)

**Comment by AaronMT** [comment]
This is needed to build on Xcode 12b6 on ```main```

File: WidgetKit/Helpers.swift [hunk 1]
```diff
@@ -13,8 +14,30 @@ var scheme: String {
...
+    _ urlSuffix: String = "",
+    query: String,
+    widgetObjectIdentifier: String,
+    widgetSize: WidgetSize
+) -> URL {
     let urlString = "\(scheme)://\(query)\(urlSuffix)"
+    
+    let prefs = BrowserProfile(localName: "profile").prefs
+    let prefKeySyncEvents = "widgetkit.telemetry.event"
+    var events = prefs.arrayForKey(prefKeySyncEvents) as? [Data] ?? []
```
Comment (garvankeeley) [inline WidgetKit/Helpers.swift:27]: `if events.count > 20 { log to sentry that we are piling up too many events and return url }`

File: Client/Application/AppDelegate.swift [hunk 2]
```diff
@@ -284,6 +284,23 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
             }
             TelemetryWrapper.recordEvent(category: .appExtensionAction, method: .applicationOpenUrl, object: object)
         }
+        
+        // TODO: Not sure where this constant for array key should live?
+        if let profile = profile, let pickledEvents = profile.prefs.arrayForKey("widgetkit.telemetry.event") as? [Data] {
+
```
Comment (garvankeeley) [inline Client/Application/AppDelegate.swift:290]: missing on this line: clearing the array

**Review by garvankeeley — CHANGES_REQUESTED** [review]

**Pull request closed without merging**